### ISSUE TITLE
Add elm-ui to the application

### ISSUE THREAD

**Issue by SimonLab** [dwyl/app-mvp-elm#22]
The current UI of the application is very basic:
![image](https://user-images.githubusercontent.com/6057298/77414018-3c390000-6db8-11ea-8c56-5c2f7787cf60.png)

Can we start using elm-ui to implement some UI parts of the application?

see also https://github.com/dwyl/learn-elm/tree/master/tutorials/elm-ui

**Comment by SimonLab** [comment]
I've tested first how to add elm-ui element to the current elm html structure without rewritting everything.

The `main` view is return the type `Browser.Document Msg`:
https://github.com/dwyl/app-mvp-elm/blob/e21abaeb29b1ff487bd559298f39e9a82bc0a250/src/Main.elm#L213

The `Browser.Document` is defined as:
```elm
type alias Document msg =
    { title : String
    , body : List (Html msg)
    }
```

We see that the `body` is a list of `Html Msg`.

To render the `elm-ui` elements we can use the `layout` function which returns `Html msg`
![image](https://user-images.githubusercontent.com/6057298/77539658-a8d7fb80-6e99-11ea-8945-791a6fe5b863.png)

This means that we can for example create a simple function:
```elm
hello : Html Msg
hello =
    Element.layout [] <| Element.el [ Element.height Element.shrink, Element.centerX ] (Element.text "hello")
```

That we can use in a `Browser.Document`

```elm
view : Model -> Page.PageStructure Msg
view model =
    { title = "Hello"
    , content =
        [ ...
        , hello
        , ...
        ]
    }
```

**Comment by SimonLab** [comment]
To Add events to elm-ui element we need to replace the functions from `Html.Events` module by the one from [Element.Events](https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element-Events):
![image](https://user-images.githubusercontent.com/6057298/77551913-cb264500-6eaa-11ea-8f43-e37037385e1b.png)

However we can see that some events are not defined in this module, for example when the text of an input change, i.e. `onInput` event is not defined here.

Elm-ui define these types of events which are directly linked to an html element (button, input...) directly in the definition of these element. For example if we want to use an input text, the module [Element.Input](https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element-Input) define the function `text`:
![image](https://user-images.githubusercontent.com/6057298/77552209-2821fb00-6eab-11ea-9d79-c29218b84672.png)

And we can see that this function takes a record as a second parameter where we can defined the message for `onChange`

**Comment by nelsonic** [comment]
@SimonLab yeah, I'm not a huge fan of this, but they appear to directly mapped. 👍

**Comment by SimonLab** [comment]
So I've manage to recreate the structure for the home page with elm-ui:
![image](https://user-images.githubusercontent.com/6057298/77558703-2825f900-6eb3-11ea-83b3-deb34b4078fc.png)

I'm now looking at how to make this nicer. The way we've done that so far is by using Tachyons css classes. However the idea with elm-ui is to avoid using html and css.
The idea would be to create ui elements composing the functions from [Element.Font](https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element-Font).

**Comment by SimonLab** [comment]
I've been switching the pages from "normal" Elm html to elm-ui.
The only page remaining is the capture page which display all the captures and timers.
I'm more specifically working on the timer buttons. Depending on the state of the capture (e.g. in progress or completed) we want the timer button to be disabled.
elm-ui however doesn't provide the `disabled` attribute for accessibility reasons:
https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element-Input#Label
![image](https://user-images.githubusercontent.com/6057298/77646259-14829d00-6f5c-11ea-88ad-d34b7828216b.png)
![image](https://user-images.githubusercontent.com/6057298/77646266-1b111480-6f5c-11ea-85ff-aef1cd924584.png)

I found interesting that elm-ui force you to rethink how to create UI. Having a limited number of functions (compare to the number of html tags/attributes) force you to think about the content and simplify the way you build the interface, the same way Elm limit the number of function/syntax compare to JavaScript

**Comment by SimonLab** [comment]
Capture page updated and using elm-ui:
![image](https://user-images.githubusercontent.com/6057298/77652036-8ca19080-6f65-11ea-957b-c35bd35f7bbc.png)

Some improvement to add on the alignments and buttons but the structure is here.

**Comment by nelsonic** [comment]
Beautiful! Ship!! :shipit: (😉keep up the great progress!)

**Comment by SimonLab** [comment]
A bit better:
![image](https://user-images.githubusercontent.com/6057298/77672082-5cb3b680-6f80-11ea-8496-a780a191e35e.png)

One question we can ask is how to reuse elm-ui view, for example at the moment the dwyl logo is displayed and defined on all the pages. Instead we could create a specific module which encapsulate the UI logo definition.

This is the idea explained here: https://github.com/rofrol/elm-ui-cookbook#structuring-code
![image](https://user-images.githubusercontent.com/6057298/77672356-b61be580-6f80-11ea-88d5-649c4c0976c4.png)

**Comment by SimonLab** [comment]
Some aspects of the UI that can be improved:
- [x] Logo seems a bit too big and takes a bit too much screen space. At the moment it is uses as a home button: reduce the size of the image
- [x] Remove "DWYL Application" as it is not an essential piece of information
- [x] The "Add capture" button doesn't contrast with the background. Create an obvious Call To Action by adding colours. Do we also want a different UI when the user hover the button?
- [x] Use spaces to distinguish even more between the "add capture" and "list of captures" section. We could also use a different background colour for the two sections.
- [x] Keep size consistency between the timer buttons: Create a fix size for the button
- [x] Use different colours for the status of the button timers, e.g gree to start, red to stop, another green for completed?
- [ ] There is not enough seperation between the capture
  - [ ] Create a background colour for the all capture row
  - [x] Add more space between the captures
  - [x] Use a line separation between the captures (e.g. hr html tag)
- [x] The time information should contrast more compare to the capture text
  - [x] use a different font family size and style

I'm also going to watch the following video (the tutorial is using Scrimba):
[![image](https://user-images.githubusercontent.com/6057298/77773124-11ada800-7041-11ea-8a52-c7806fb73628.png)](https://www.youtube.com/watch?v=tRpoI6vkqLs)

**Comment by SimonLab** [comment]
To reduce the image for the logo, I first tried to specify the `height` for the logo section compare to the space for the content section using [fillPortion](https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element#fillPortion) function. This function allow you to specify which amount of screen space each section should have:

>So, two elements, one with width (fillPortion 2) and one with width (fillPortion 3). The first would get 2 portions of the available space, while the second would get 3.

However I haven't managed to reduce the space of the logo with this method.
I've finally used the `px` function to specify the widht/height pixel size, e.g. `width (px 150)`. Then using spacing and padding to add space around the logo.

`spacing` define the space between the child elements, in our case between the logo and content section.
`padding` define the space between the content and the out layer of the content:

![image](https://user-images.githubusercontent.com/6057298/77905496-422e5580-727e-11ea-80b6-963f13ea6233.png)

![image](https://user-images.githubusercontent.com/6057298/77905549-57a37f80-727e-11ea-90b2-28b1179ba749.png)

**Comment by nelsonic** [comment]
@SimonLab consider just re-sizing the image and uploading a smaller one. It saves bandwidth for everyone. 😉

**Comment by SimonLab** [comment]
- [x] Need to wrap the text of the capture

Currently trying to figure out capture display:
![image](https://user-images.githubusercontent.com/6057298/77912783-16fe3300-728b-11ea-9d70-195921ad4ee9.png)



Using `paragraph` to wrap text works when the label for the checkboxes have a fix width, here 500px:
![image](https://user-images.githubusercontent.com/6057298/77916592-326c3c80-7291-11ea-89f6-caa629c3c82c.png)

![image](https://user-images.githubusercontent.com/6057298/77915427-5dee2780-728f-11ea-86ad-ffb4cbb4bed9.png)

I've manage to make the paragraph wrap. I needed to make sure that the child element in the first part were all defined with `width fill` to make them take as much space available

**Comment by SimonLab** [comment]
- [ ] Search how to make the timer section width fixed

I'm not sure why the width of the timer is changing as I thought that the width were fixed when all the elements of the row were defined with `width fill`. This should divide the space for the three sections evenly. We can see when the timer text is changing the spaces of the contents are changing too: 

![width-fill](https://user-images.githubusercontent.com/6057298/77920865-e8865500-7296-11ea-8306-7f8cf29cae5a.gif)

**Comment by SimonLab** [comment]
I have updated some colours for the buttons:
![image](https://user-images.githubusercontent.com/6057298/77927461-28513a80-729f-11ea-8e49-962f722f6ffc.png)

**Comment by nelsonic** [comment]
@SimonLab looks good. 🙌

**Comment by SimonLab** [comment]
Adding line separation between captures:
![image](https://user-images.githubusercontent.com/6057298/77931868-acf28780-72a4-11ea-872b-9a7a09f09b57.png)

**Comment by SimonLab** [comment]
I'm currently updating the pages remaining:
- [x] "home" page
![image](https://user-images.githubusercontent.com/6057298/78026920-5775b380-7354-11ea-82de-d870cf189594.png)

- [x] Timers page
![image](https://user-images.githubusercontent.com/6057298/78026990-7b38f980-7354-11ea-9017-45150262d49d.png)

- [x] Make links more obvious by using the teal dwyl colour

**Comment by SimonLab** [comment]
Inspired by how [tachyons](https://tachyons.io/components/buttons/basic/index.html#0) manage hover/focus on buttons, I've added a transition on the buttons which reduce the opacity:

Create a `dim` class in `main.css`:
```css
.dim {
    opacity: 1;
    transition: opacity .15s ease-in;
}

.dim:hover, .dim:focus {
    opacity: .5;
    transition: opacity .15s ease-in;
}

.dim:active {
    opacity: .8;
    transition: opacity .15s ease-out;
}
```

I don't think elm-ui has a function which manages the opacity of an element so I'm using the `htmlAttribute` function to add the `dim` class:

```elm
buttonAttrs : List (Elt.Attribute msg)
buttonAttrs =
    [ Elt.centerX
    , EltBrd.solid
    , EltBrd.width 1
    , EltBrd.rounded 3
    , Elt.padding 5
    , Elt.htmlAttribute <| HtmlAttr.class "dim"
    ]
```

![dim-button](https://user-images.githubusercontent.com/6057298/78342797-bd02b380-7591-11ea-9118-de5b6239b8a1.gif)

**Comment by SimonLab** [comment]
I was looking to underline the timers on hover using the [focused](https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element#focused) and the [underline](https://package.elm-lang.org/packages/mdgriffith/elm-ui/latest/Element-Font#underline) function

`focused : List Decoration -> Attribute msg`
`underline : Attribute msg`

Unfortunately we can see that `focused` takes a list of `Decoration` but `underline` returns an `Attribute msg` and from the discussion I've seen on the elm-ui channel on slack, it doesn't seem possible to use underline with focused at the moment. So I'll use `htmlAttribute` to add an `underline` class to the timer link.

![underline-link](https://user-images.githubusercontent.com/6057298/78353223-9ba9c380-75a1-11ea-874f-62151462a216.gif)

**Comment by nelsonic** [comment]
@SimonLab showing visual feedback on buttons is a proactive initiative. 🙌  
_However_, hovering over a button it should bring it into focus, 
i.e. **_increase_ contrast** from the white background not _reduce_ it. 
The _last_ thing we want to do is _reduce_ the contrast of the button making it _less_ legible when someone is hovering over it. The UX of "greying out" something by reducing opacity indicates the button is "disabled", i.e. the action cannot be performed. 

![buttons-opacity-reduced-when-hovering](https://user-images.githubusercontent.com/194400/78350502-3b188780-759d-11ea-83b0-4851e7644870.gif)

In this CSS Tricks Tutorial [Chris Coyier](https://github.com/chriscoyier) shows how to reduce the opacity of _other_ buttons (_not being hovered over_) to show the one that _is_ in focus: https://css-tricks.com/hover-on-everything-but
i.e. the button that is hovered over is brought into focus and the _other_ buttons have reduced opacity.

Posted a question to UX to ask what the ideal micro interaction for hovering over a button is:
https://ux.stackexchange.com/questions/132362/what-micro-interaction-to-use-when-hovering-over-a-button

> Please don't take this as personal criticism. I just want to _understand_ what the best UX for buttons are so we can build the best UX for the people using the App. 👍

**Comment by SimonLab** [comment]
Thanks for testing the hover effect @nelsonic.
Agree that the contract should be more obvious on hover instead of having this "dim" effect.
I'm wondering if reducing the opacity on all the elements that are not focused might be a bit too much in the opposite way :thinking:

**Comment by nelsonic** [comment]
@SimonLab yeah, I genuinely don't know what the ideal interaction is on buttons. 
In the past I've added a drop-shadow to buttons to show they are clickable.
See: https://tailwindcss.com/docs/box-shadow/#focus-outline-shadow
